File: 238-product-of-array-except-self/product-of-array-except-self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut ans = vec![1; n];
        let mut curr = 1;
        for i in 0..n {
            ans[i] *= curr;
            curr *= nums[i];
        }
        curr = 1;
        for i in (0..n).rev() { 
            ans[i] *= curr;
            curr *= nums[i];
        }
        ans
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int zc = 0, prod = 1;
        int[] ans = new int[n];
        for(int i : nums)
        {
            if(i != 0) prod *= i;
            else zc++;
        }
        if(zc > 1) return ans;
        for(int i = 0; i < n; i++)
        {
            if(zc > 0) ans[i] = (nums[i] == 0) ? prod : 0;
            else ans[i] = prod / nums[i];
        }
        return ans;
    }
}